(bugfix) Use a defined __hash__() for unhashable middleware:

fselmo opened the pull request:
### What was wrong?

- Closes #3461
- There is a bug where the middleware builders, which return curry-wrapped ``build`` functions, end up being given the same identifier when a name is not provided for the middleware in the onion.

### How was it fixed?

This fixes the issue by defining a unique hash for the base ``Web3Middleware`` class and using that hash in the case where the middleware is unhashable and therefore indistinguishable from another "built" middleware. This also uses the `__hash__()` by default for any of the class-based middleware if a name isn't provided, rather than using a `repr()` as before - which is not unique enough, leading to the complications.

### Todo:

- [x] Clean up commit history
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

![](https://www.arrowexterminators.com/assets/istock-1032959962_gvZN2oZ.jpg)

fselmo committed:
(bugfix) Use a defined __hash__() for unhashable, unnamed middleware:

Fixes a bug where the middleware which are not hashable, classes or the
``build`` method of builder classes, end up being given the same
identifier when a name is not provided for the middleware in the onion.
This fixes the issue by defining a unique hash for the base
``Web3Middleware`` class and using that hash in the case where the
middleware is unhashable and therefore indistinguishable from another
middleware.

fselmo committed:
Add middleware tests for __eq__ and test for unique key gen.

kclowes commented on tests/core/middleware/test_middleware.py:
```diff
@@ -0,0 +1,63 @@
+    assert mw1w3_a != mw1w3_b
+
+
+def test_unnamed_middleware_are_given_unique_keys(w3):
```
Was this happening when the middleware got passed in as a class too, rather than the build method? Do you think it's worth adding a test for that case?

fselmo replied:
I definitely think it's worth it to add that, will do.

kclowes commented on web3/datastructures.py:
```diff
@@ -248,15 +248,25 @@ def replace(self, old: TKey, new: TKey) -> TValue:
 
-    def _repr_if_not_hashable(self, value: TKey) -> TKey:
+    @staticmethod
+    def _build_tkey(value: TKey) -> TKey:
```
what do you think about calling this function something like `_build_name`?

fselmo replied:
That works 👍🏼

kclowes reviewed (approved):
I left a comment about naming, and adding another test if appropriate, but feel free to take or leave. Thanks for getting this in!

fselmo committed:
Changes from comments on PR #3463:

- ``_build_tkey()`` -> ``_build_name()``
- Add test for unnamed ``Web3Middleware`` added to onion, not just ``Web3MiddlewareBuilder.build()`` (functools.curry) references.

pacrob reviewed (approved):
:sheep:
